nixos/modules/services/web-apps/porn-vault/default.nix

```diff
@@ -0,0 +1,117 @@
+{
+  config,
+  pkgs,
+  lib,
+  ...
+}:
+
+let
+  cfg = config.services.porn-vault;
+  configFormat = pkgs.formats.json { };
+  defaultConfig = import ./default-config.nix;
+  inherit (lib)
+    mkIf
+    mkEnableOption
+    mkPackageOption
+    mkOption
+    getExe
+    literalExpression
+    types
+    ;
+in
+{
+  options = {
+    services.porn-vault = {
+      enable = lib.mkEnableOption "Porn-Vault";
+
+      package = lib.mkPackageOption pkgs "porn-vault" { };
+
+      autoStart = lib.mkOption {
+        type = lib.types.bool;
+        default = true;
+        description = ''
+          Whether to start porn-vault automatically.
+        '';
+      };
+
+      port = lib.mkOption {
+        type = lib.types.port;
+        default = 3000;
+        description = ''
+          Which port Porn-Vault will use.
+        '';
+      };
+
+      openFirewall = lib.mkOption {
+        type = lib.types.bool;
+        default = false;
+        description = ''
+          Whether to open the Porn-Vault port in the firewall.
+        '';
+      };
+
+      settings = {
+        enable = mkOption {
+          description = "Enable Porn-Vault configuration via Nix";
+          default = true;
+        };
+
+        settings = mkOption {
+          type = configFormat.type;
+          description = ''
+            Configuration for Porn-Vault. The attributes are serialized to JSON in config.json.
+
+            See https://gitlab.com/porn-vault/porn-vault/-/blob/dev/config.example.json
+          '';
+          default = defaultConfig;
+          apply = lib.recursiveUpdate defaultConfig;
+        };
+      };
+    };
+  };
```

Comment:
I get why this would be an option, but it's pretty abnormal in Nixpkgs -- though, I'm not sure how this would usually be handled either. I only know `settings.settings` Isn't Great

Maybe split this into `services.porn-vault.manageSettings` and `services.porn-vault.settings`?

But at the same time, having this kind of setting at all can lead to easy breakage (since as we've already discussed, porn-vault *really* needs it's config file to exist), so maybe this file should always be managed when using the module? Would be best to have some discussion here IMO

Reply:
I can remove the option to disable it if you really want me to do so, but I do believe it would be useful to support disabling it. For instance, one could disable it and then do

```nix
environment.etc."porn-vault/config.json".source = fetchUrl {
  url = some-domain/some-file.json;
  hash = whatever;
}};
```

to grab a config file from elsewhere, if for whatever reason they have it stored somewhere else.

Or they could be tweaking their config (personally I never really messed much with the config, but I do know that PV has some plugin options and stuff so I imagine it’s possible for one to need to do some tweaking) and stuff, in which case it could be annoying to rebuild the system every time. This way they could manually provide the config in /etc and then once they finish nixing it rebuild with that enabled. Tho PV config is mostly static I think, so it wouldn’t be that much of a problem either.

And the user could always import a json in the settings if they don’t want at all to nix it. 

Yeah actually I think the fetchurl thing would be entirely possible inside the settings key too, they wouldn’t need to disable the module-managed settings for that.

Yeah actually I think you’re right, I could remove the settings.enable thing. I’ll do that later today.

Reply:
Done